### PR TITLE
Create a RSS or ATOM feed for news page, fixes #200

### LINKED ISSUE #200: Create a RSS or ATOM feed for news page
Page is static now but we could put all the static info in an array and once the MVC structure is set up generate a web feed automatically.

### PR DESCRIPTION
**Pull request by TheoChevalier**
Link to RSS is included in <head>, not sure if we want a direct link in the /news view.

Also, I have no idea why php-cs-fixer fixes code that previous commit should already have fixed (I don’t think it got updated, but that may be that)

Note that links for authors could be anything, pages on mozillians.org, websites, or just plain text.

### CODE REVIEW

File: app/models/changelog.php [hunk 1]
```diff
@@ -66,19 +32,49 @@
             break;
     }
 
-    return "<span class=\"release_tag {$type}\">{$text}</span> ";
+    return "<span class=\"release_tag {$type}\">{$text}</span>";
+};
+
+$get_sections = function ($var) {
```
Comment (flodolo) [inline app/models/changelog.php]: Use something less generic than `$var`?

**Comment by flodolo** [comment]
RE php-cs changes, it's really weird. I have an open PR and php-cs doesn't catch those changes.

File: app/models/changelog.php [hunk 2]
```diff
@@ -114,3 +108,27 @@
 
     return $link;
 };
+
+// Helper to generate change author names
```
Comment (flodolo) [inline app/models/changelog.php]: Not sure what you wanted to tell here. Maybe a link to the GitHub account makes more sense? If there's code, there's going to be a GitHub account.

Comment (TheoChevalier) [inline app/models/changelog.php]: Yeah, and Travis doesn’t seem to complain. We’re not testing PHP 7 on Travis because that’s what we’re all using locally I guess?

**Comment by TheoChevalier** [comment]
Yeah, and Travis doesn’t seem to complain. We’re not testing PHP 7 on Travis because that’s what we’re all using locally I guess?

File: app/views/changelog_rss.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+$output .= '  <atom:link href="/rss" rel="self" type="application/rss+xml" />' . "\n";
+$output .= '  <title>Transvision</title>' . "\n";
+$output .= '  <description>Transvision Release Notes</description>' . "\n";
+$output .= '  <lastBuildDate>' . date(DATE_RSS) . '</lastBuildDate>' . "\n";
+$output .= '  <link>https://'  . '</link>' . "\n";
+
+foreach ($changelog as $release => $changes) {
+    $output .= '  <item>' . "\n";
+    $output .= '  <guid isPermaLink="false">' . sha1($release . date('W')) . '</guid>' . "\n";
+    $output .= '  <pubDate>' . date(DATE_RSS) . '</pubDate>' . "\n";
```
Comment (flodolo) [inline app/views/changelog_rss.php]: Should we use the date of the release?

File: app/views/changelog_rss.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+        $output .= ' ' . $local_link($attributes['message'][0]);
+
+        $output .= isset($attributes['authors'])
+                                ? $authors($attributes['authors'])
+                                : '';
+    }
+    $output .= '<br>' . $github_link($release);
+    $output .= ' ]]></description>' . "\n";
+    $output .= '  <link>' . '/news/#v' . $release . '</link>' . "\n";
```
Comment (flodolo) [inline app/views/changelog_rss.php]: This link will be broken on an external reader, wouldn't it?

Comment (flodolo) [inline app/views/changelog_rss.php]: I actually end up with an empty link :grey_question:

Comment (flodolo) [inline app/views/changelog_rss.php]: Ah, that's because the link should before the `<item>`, I don't we need a link on each item either.

File: app/views/changelog_rss.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+$output  = header('Content-type: application/xml; charset=UTF-8');
+$output .= header('Access-Control-Allow-Origin: *');
+$output .= '<?xml version="1.0" encoding="utf-8"?>' . "\n";
+$output .= '<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom">' . "\n";
+$output .= ' <channel>' . "\n";
+$output .= '  <atom:link href="/rss" rel="self" type="application/rss+xml" />' . "\n";
+$output .= '  <title>Transvision</title>' . "\n";
+$output .= '  <description>Transvision Release Notes</description>' . "\n";
+$output .= '  <lastBuildDate>' . date(DATE_RSS) . '</lastBuildDate>' . "\n";
+$output .= '  <link>https://'  . '</link>' . "\n";
```
Comment (flodolo) [inline app/views/changelog_rss.php]: The empty links come from here.

File: app/models/changelog.php [hunk 6]
```diff
@@ -114,3 +108,27 @@
...
+$authors = function ($list) {
+    $text = ' (';
```
Comment (flodolo) [inline app/models/changelog.php]: Maybe using an array would be more readable? Example
```
$author_urls = [
    'flod'       => 'https://twitter.com/flod',
    'pascal'     => 'https://twitter.com/pascalchevrel',
    'tchevalier' => 'https://twitter.com/t_chevalier',
];
$text = ' (';

foreach ($list as $author) {
    if (isset($author_urls[$author])) {
        $text .= "<a href=\"$author_urls[$author]\">{$author}</a>";
    } else {
        $text .= $author;
    }
    $text .= ($author === end($list)) ? ')' : ', ';
}

return $text;
```

File: app/models/changelog.php [hunk 7]
```diff
@@ -66,19 +32,49 @@
...
-    $keys = array_keys($releases);
-    $previous = $keys[array_search($release, $keys, true) - 1];
+    if ($release > 1) {
+        $keys = array_keys($releases);
+        $previous = $keys[array_search($release, $keys, true) - 1];
 
-    return <<<LINK
+        return <<<LINK
 <p class="github_link">See the complete list of
 <a href="https://github.com/mozfr/transvision/compare/v{$previous}...v{$release}">code changes from version {$previous} on Github</a>.
```
Comment (flodolo) [inline app/models/changelog.php]: Nit (not yours, pre-existing): GitHub

File: app/models/changelog_data.php [hunk 8]
```diff
@@ -0,0 +1,1468 @@
...
+            'authors' => ['tchevalier'],
+        ],
+        [
+            'message' => ['Add upstream remote repository automatically in setup script'],
+            'section' => ['developers'],
+            'type'    => ['better'],
+            'issues'  => ['466'],
+            'authors' => ['pascal'],
+        ],
+        [
```
Comment (flodolo) [inline app/models/changelog_data.php]: You lost a line (issue 476)

File: app/models/changelog_data.php [hunk 9]
```diff
@@ -0,0 +1,1468 @@
...
+            'authors' => ['tchevalier'],
+        ],
+        [
+            'message' => ['Regression: fix Select button on TMX download view'],
+            'section' => ['end_user'],
+            'type'    => ['bug'],
+            'authors' => ['tchevalier'],
+        ],
+        [
+            'message' => ['Use proper plural form for the string "Displaying x results"'],
```
Comment (flodolo) [inline app/models/changelog_data.php]: This is the not the right line (copy & paste duplicate, should be issue 461)

File: app/models/changelog_data.php [hunk 10]
```diff
@@ -0,0 +1,1468 @@
...
+        ],
+        [
+            'message' => ['Check if Composer (PHP dependency manager) is not installed globally before downloading it'],
+            'section' => ['developers'],
+            'type'    => ['better'],
+            'authors' => ['Julien Itard'],
+        ],
+        [
+            'message' => ['<code>Strings::endsWith()<code> and </code>Strings::startsWith()</code> now accept arrays of strings for checking, e.g. <code>Strings::startsWith(\'hello!\', [\'!\', \'?\', \'.\'])</code>'],
+            'section' => ['end_user'],
```
Comment (flodolo) [inline app/models/changelog_data.php]: Not sure if this one is causing the problem, but the first issues in 3.5 are missing the title.

**Comment by flodolo** [comment]
There are some issues with the generate HTML code: the link to GitHub and the next release title ends up inside the last `<ul>` element.

One suggestion: copy the text of the entire page in your local version, copy the text from the current release note in prod, then run a diff. It looks like there are a few differences (I noted a few).

File: app/models/changelog_data.php (hunk 10 above)
Comment (TheoChevalier) [inline app/models/changelog_data.php]: I was not reseting the section variable between two releases, so when the first section of a release equals the last section of previous release, it was not displayed

**Comment by TheoChevalier** [comment]
I did some diffs of the HTML last summer when I started this pull request, it helped catch some mistakes but was hard to follow, given the small adjustments I had to make. Will give it another look

File: app/views/changelog_rss.php [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+$output .= '  <lastBuildDate>' . date(DATE_RSS) . '</lastBuildDate>' . "\n";
+$output .= '  <link>https://transvision.mozfr.org/news/</link>' . "\n";
+
+foreach ($changelog as $release => $changes) {
+    $section = '';
+    $output .= '  <item>' . "\n";
+    $output .= '  <guid isPermaLink="false">' . sha1($release . date('W')) . '</guid>' . "\n";
+    $output .= '  <pubDate>' . date(DATE_RSS, strtotime($releases[$release])) . '</pubDate>' . "\n";
+    $output .= '  <title>' . 'Version ' . $release . ' Release Notes</title>' . "\n";
+    $output .= '  <description><![CDATA[ ';
```
Comment (flodolo) [inline app/views/changelog_rss.php]: I know I threw you off on this, but I realized we actually need the link here. If you load the RSS as Live Bookmark, it will always get to the main page.
```
$output .= "  <link>https://transvision.mozfr.org/news/#v{$release}</link>\n";
```

File: app/models/changelog.php [hunk 12]
```diff
@@ -114,3 +102,19 @@
...
     return $link;
 };
+
+// Helper to generate a link for each author
+$authors = function ($list) {
+  $text = ' (';
+
+  foreach ($list as $author) {
+      if (isset($author_urls[$author])) {
+          $text .= "<a href=\"$author_urls[$author]\">{$author}</a>";
```
Comment (flodolo) [inline app/models/changelog.php]:
```
"<a href=\"{$author_urls[$author]}\">{$author}</a>"
```

File: app/models/changelog.php [hunk 13]
```diff
@@ -114,3 +102,19 @@
 
     return $link;
 };
+
+// Helper to generate a link for each author
+$authors = function ($list) {
```
Comment (flodolo) [inline app/models/changelog.php]: `$author_urls` is not available in this function, needs `use`

**Comment by TheoChevalier** [comment]
Had to add additional logic to get the GitHub link displayed outside of </ul>, but it seem to work

**Comment by TheoChevalier** [comment]
rebased

**Comment by flodolo** [comment]
Looks great to me, r+ :+1: 

Let's see if @pascalchevrel wants to give it a check, especially considering he's the main update of release notes

File: app/views/changelog_rss.php [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    $output .= '<br>' . $github_link($release);
+    $output .= ' ]]></description>' . "\n";
+    $output .= '  </item>' . "\n";
+}
+
+$output .= ' </channel>' . "\n";
+$output .= '</rss>' . "\n";
+
+printf($output);
```
Comment (pascalchevrel) [inline app/views/changelog_rss.php]: why printf() and not just print?

File: app/views/changelog_rss.php [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    $output .= "  <link>https://transvision.mozfr.org/news/#v{$release}</link>\n";
+    $output .= '  <description><![CDATA[ ';
+    foreach ($changes as $change => $attributes) {
+        if ($section != $attributes['section'][0]) {
+            $section = $attributes['section'][0];
+            $output .= '<br><h3>' . $get_sections($section) . '</h3>';
+        }
+        $output .= '<br>';
+        $output .= isset($attributes['type'])
+                                ? '[' . $relnotes($attributes['type'][0]) . ']'
```
Comment (pascalchevrel) [inline app/views/changelog_rss.php]: Any reason to indent the condition statements so much?

File: app/views/changelog_rss.php [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    $output .= '  <pubDate>' . date(DATE_RSS, strtotime($releases[$release])) . '</pubDate>' . "\n";
+    $output .= '  <title>' . 'Version ' . $release . ' Release Notes</title>' . "\n";
+    $output .= "  <link>https://transvision.mozfr.org/news/#v{$release}</link>\n";
+    $output .= '  <description><![CDATA[ ';
+    foreach ($changes as $change => $attributes) {
+        if ($section != $attributes['section'][0]) {
+            $section = $attributes['section'][0];
+            $output .= '<br><h3>' . $get_sections($section) . '</h3>';
+        }
+        $output .= '<br>';
```
Comment (pascalchevrel) [inline app/views/changelog_rss.php]: Whu do you need to insert line breaks in the RSS feed?

File: app/views/changelog_rss.php (hunk 15 above)
Comment (TheoChevalier) [inline app/views/changelog_rss.php]: Not really :) Personnaly I find it less cluttured that way, but we can move ithem below the isset I guess

File: app/views/changelog_rss.php (hunk 16 above)
Comment (TheoChevalier) [inline app/views/changelog_rss.php]: They are still interpreted as line breaks in the feed, so if I remove them it would look very compact. Do you think the feed should be even more minimalist?

File: app/controllers/changelog.php [hunk 17]
```diff
@@ -2,4 +2,9 @@
 namespace Transvision;
 
 include MODELS . 'changelog.php';
-include VIEWS . 'changelog.php';
+
+if ($page == 'rss') {
+    include VIEWS . 'changelog_rss.php';
+} else {
+    include VIEWS . 'changelog.php';
+}
```
Comment (pascalchevrel) [inline app/controllers/changelog.php]: suggestion, shorten that with a ternary operator:
```php
include VIEWS . 'changelog' . ($page == 'rss' ? '_rss' : '') . '.php';
```

**Comment by pascalchevrel** [comment]
LGTM with a few nits

File: app/views/changelog_rss.php (hunk 16 above)
Comment (pascalchevrel) [inline app/views/changelog_rss.php]: It doesn't look very compact to me in Firefox, I am actually more worried about how feed readers will interpret the line break

**Comment by pascalchevrel** [comment]
in https://github.com/TheoChevalier/transvision/blob/rss-feed-relnotes/app/models/changelog_data.php#L831

hred should be href (was wrong in the current release notes)

**Comment by pascalchevrel** [comment]
The feed doesn't validate with https://validator.w3.org/feed/#validate_by_input

**Comment by TheoChevalier** [comment]
Fixed nits, feeed now validates
